docs/architecture.md

```diff
@@ -0,0 +1,225 @@
+# Loki's Architecture
+
+This document will expand on the information detailed in the [Loki
+Overview](overview/README.md).
+
+## Multi Tenancy
+
+All data both in memory and in long-term storage is partitioned by a tenant ID,
+pulled from the `X-Scope-OrgID` header from the request when Loki is ran in
+multi-tenant mode. When Loki is **not** in multi-tenant mode, the header
+is ignored and the tenant ID is set to "fake", which will appear in the index
+and in stored chunks.
+
+## Modes of Operation
+
+![modes_diagram](modes_of_operation.png)
+
+Loki has a set of components (defined below in [Components](#components)) which
+are internally referred to as modules. Each component spawns a gRPC server for
+internal traffic and an HTTP/1 server for external API requests. All components
+come with an HTTP/1 server, but most only expose readiness and health endpoints.
+
+While each component can be ran in a separate process, Loki also supports running
+all components in a single process. Running all components in a single process is
+referred to as "single process" mode, while the other mode is the "horizontally"
+scalable mode.
+
+When Loki runs in single process mode, individual components continue to
+communicate to one another over gRPC using the gRPC listen port of the overall
+process.
+
+Single process mode is ideally suited for localy development, small workloads,
+and for evaluation purposes. Single-process mode can be scaled with multiple
+processes with the following limitations:
+
+1. Local index and local storage cannot currently be used as they are not safe
+   for concurrent access by multiple processes.
+2. Individual components cannot be scaled independently, so it is not possible
+   to have more read components than write components.
+
+## Components
+
+### Distributor
+
+The **distributor** service is responsible for handling incoming streams by
+clients. It's the first stop in the write path for log data. Once the
+distributor receives a set of streams, they are split into batches and sent to
+multiple [ingesters](#ingester) in parallel.
```

Comment:
Distributor also does some validation of the incoming data, things like # of labels.  check out validations.go from the cortex project